libraries/architecture/src/main/kotlin/io/element/android/libraries/architecture/Async.kt

```diff
@@ -18,51 +18,119 @@ package io.element.android.libraries.architecture
 
 import androidx.compose.runtime.MutableState
 import androidx.compose.runtime.Stable
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.InvocationKind
+import kotlin.contracts.contract
 
+/**
+ * Sealed type that allows to model an asynchronous operation.
+ */
 @Stable
 sealed interface Async<out T> {
+    data class Failure<out T>(
+        val exception: Throwable,
+        val prevData: T? = null,
```

Comment:
I'd add a comment here about what `prevData` is used for.

Reply:
Added docstrings